Improved AI BaseBuilder

reaperrr opened the pull request:
This improves 3 things:

**1. The low power check.** 
Previously, it would check whether players' `ExcessPower` was less than `MinimumExcessPower` or whether the sum of all `Power` traits on the structure it wanted to build _exceeded_ the players' `ExcessPower`, and if either was true, build a power plant instead.
The catch: power _draw_ is represented by _negative_ numbers, so the only structures that would trigger the 2nd check were - power plants.
In other words, as long as current `ExcessPower` was equal or higher than `MinimumExcessPower`, the whole check would effectively fail, hence why the AI was out of power about half the time before I introduced `MinimumExcessPower` recently.
Now the 2nd check properly evaluates whether current `ExcessPower` + to-build actors' `Power` would be _lower_ than `MinimumExcessPower`, and in that case build a power plant instead.
Additionally, unlike before, it now also performs the "do I have enough power" check for the higher-priority overrides (refineries, production buildings, silos).
Last but not least, it now also gives powerplants priority if already on low power (no idea why it was different before).

**2. Repeated placement failures.**
Previously, the AI would constantly and stubbornly retry to build and place a structure even if out of space.
Now, after 3 consecutive failed attempts to place a structure, it is assumed that the AI has run out of build space and further production of structures is paused for a minute (default value of `StructureProductionResumeDelay`) and only resumed if the number of buildings decreased (or number of construction yards increased) in the meantime.

**3. Placement failure of naval structures.**
Previously, the AI would try to build naval structures even if there was no water within the buildable area. Now it first checks whether there is enough water in the base perimeter, and disables production of naval buildings if not. 
If there is actually such water space, when deciding whether to build a naval structure it checks if a building providing buildable area is close enough to that water. Only if that condition is met as well, the naval structure will be produced.

In summary, this should considerably reduce lag spikes from repeated building placement failures (particularly if it runs out of space) and finally prevent the AI reliably from driving itself into low power.

Fixes #8116, mostly adresses #4717.

reaperrr committed:
Fixed AI BaseBuilder low power check.

This was extremely borked:
The priority overrides for refinery, production and silo ignored power completely and never checked whether the structure might drive the AI into low power state; whereas the check for regular building checked whether the structure's power -exceeded- MinimumExcessPower (previously just 0). The catch is that power draw is represented by negative numbers, so the only buildings that would trigger this were - power plants.
Now it checks whether the sum of building power draw (negative) and AI's current power level are lower than MinimumExcessPower, if not, build the structure, if yes, build a power plant instead.
Additionally, this check is now performed for the early-game priority overrides as well.

Last but not least the AI would not(!) give power plants priority if it was already low on power. This has been fixed as well.

reaperrr committed:
Check for consecutive failures to place a structure.

After 3 consecutive failed attempts, wait one minute before resetting the counter and trying again.

reaperrr committed:
Avoid consecutive attempts to build naval structures

If not enough water space can be found inside the base perimeter, stop the AI from trying to build naval production buildings permanently until it deploys another construction yard.

If enough water is available within the base perimeter, check whether any
building that provides buildable area (for adjacency) is close enough to
water, otherwise don't even start producing this naval structure.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -153,33 +167,92 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
 				}
+
+				if (power != null && refinery != null
+					&& (refinery.Traits.WithInterface<PowerInfo>().Where(i => i.UpgradeMinEnabledLevel < 1).Sum(p => p.Amount) + playerPower.ExcessPower)
```
This check is done a bunch.  Could you pull it out into a helper function?

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,10 +40,19 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 
 		public void Tick()
 		{
+			// If failed to place something 3 consecutive times, wait 1 minute until retrying to build something
```
Not necessarily 1 minute (depends on yaml).

pchote replied:
Also, 3 is a rather magic number.  Can we pull that out into a parameter too?

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,10 +40,19 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 
 		public void Tick()
 		{
+			// If failed to place something N consecutive times, wait N ticks until resuming building production
```
Nit: that second N should be an M (it is a different arbitrary number to the first one).

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,10 +40,19 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 		{
+			// If failed to place something N consecutive times, wait N ticks until resuming building production
+			if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks > 0)
+				--failRetryTicks;
```
Do you really intend on decrementing `failRetryTicks` twice per tick?  You probably want to drop the `--` inside the if check.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,10 +40,19 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+			// If failed to place something N consecutive times, wait N ticks until resuming building production
+			if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks > 0)
+				--failRetryTicks;
+			else if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks <= 0)
```
and here too.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -130,22 +145,31 @@ ActorInfo GetProducibleBuilding(string commonName, IEnumerable<ActorInfo> builda
 			return available.RandomOrDefault(ai.Random);
 		}
 
+		bool SufficientPowerForActor(ActorInfo actorInfo)
```
`HasSufficientPowerForActor`?

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -153,33 +177,76 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
+						return navalproduction;
+					}
+					else
+						waterAvailable = false;
```
If you haven't expanded to the full range of your base radius then it may be possible to place a naval yard later even if you can't place one right now.

pchote replied:
A more accurate way to test this would be to count the number of water cells within the current available base ranges.  If it's < 5 (or some other arbitrary number that should be large enough to justify building a naval yard) then turn it off.  Turn it back on and recalculate if new bases become available.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -199,24 +266,34 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
+					&& ai.Info.BuildingCommonNames["NavalProduction"].Contains(name))
+					continue;
+
+				if (waterAvailable && ai.Info.BuildingCommonNames.ContainsKey("NavalProduction")
```
This is mostly duplicating the check from above - how about pulling it out into a temporary variable above?

obrakmann commented:
Haven't looked at the code yet, only played two test games. In both games, I noticed that several AIs would either not produce any buildings at all, or stop relatively early in the tech tree (they only had power, a refinery and a silo).

The replays are [here](https://seacloud.cc/d/5d1adbebce/files/?p=/OpenRA-2015-07-22T130350Z.orarep) and [here](https://seacloud.cc/d/5d1adbebce/files/?p=/OpenRA-2015-07-22T140921Z.orarep). 

But the low power issue seems to be gone indeed. Did not see that at all in the replays.

reaperrr commented:
OK, I suspect the last chunk with the `canPlaceNavalProduction` check is bollocks and blocks production, I'll test that first.

clemty commented:
>  if no space is found for the naval structure, disables construction of naval structures permanently for the rest of that game.
wouldn't that be overkill in case the AI would somehow acquire another construction yard near water? (that might not be the case now, but what if AI gets good enough to spawn a second base sometime?)

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +42,44 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 
 		public void Tick()
 		{
+			// If failed to place something N consecutive times, wait M ticks until resuming building production
```
This can be simplified to
```csharp
// If failed to place something N consecutive times, wait M ticks until resuming building production
if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks <= 0)
    failCount = 0;

// Only update once per second or so
```

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +42,44 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+				checkedWater = true;
+			}
+
+			if (checkedWater && !waterAvailable)
```
You set checkedWater, then check if you should unset it.  This logic is twistier than it needs to be.

pchote replied:
Since you're doing this check in Tick anyway, you might as well have it check against the currently available build area, and reset it each time a building is placed or lost.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -84,13 +120,26 @@ bool TickQueue(ProductionQueue queue)
+				var navalBuilding = ai.Info.BuildingCommonNames.ContainsKey("NavalProduction")
+					&& ai.Info.BuildingCommonNames["NavalProduction"].Contains(currentBuilding.Item);
+
+				// Don't count naval buildings towards regular failCount
```
You could then remove this special case hack.

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -153,33 +211,69 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
+				}
+			}
+
+			// Make sure that we can can spend as fast as we are earning
```
The comment doesn't match what this is doing.  You can increase spending by building any new production structure - there's nothing special about naval.

penev92 replied:
Also "can can".

reaperrr commented:
Updated.

obrakmann commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -231,6 +238,30 @@ public void Activate(Player p)
+				var countWaterTiles = Map.FindTilesInCircle(b.Location, Info.MaxBaseRadius)
+					.Where(c => Player.World.Map.Contains(c)
+						&& Player.World.Map.GetTerrainInfo(c).Type == "Water")
+					.Count();
```
Shouldn't this check if the water is in buildable range, too (`BuildingInfo.Adjacent`)? Otherwise the AI would still try to build naval structures even before it could actually place them just because there is water somewhere within the base radius. Not sure, really. Maybe this is good enough.

obrakmann commented:
Works nicely now, otherwise.

reaperrr commented:
Updated, refactored the naval check once more.

Rather than just counting an arbitrary number of water cells, it now checks whether there is a least one water cell where all 8 adjacent cells are water cells, too.

Additionally, if the above is the case, when deciding what structure to build next it then checks whether such a cell is within a radius of N (defaults to 8 since that's the Adjacent value of RA naval buildings) around any current building providing buildable area, and only starts production of a naval building if this is the case as well.

This should be far more robust now, but needs some testing, of course.

reaperrr commented:
There's one more small thing I'd like to add, so don't merge this yet even if everything works fine.

reaperrr committed:
Further improved building placement failure check

Since naval structures have their own safety measures now and therefore shouldn't count towards failCount under normal circumstances, we can now assume that 3 consecutive placement failures mean lack of space. Therefore, rather than unconditionally resetting the failCount and retry every N ticks, we now cache the number of buildings and construction yards at the time of the 3rd consecutive failure and if the number of buildings hasn't decreased and number of construction yards not increased, we assume there is still not enough space and reset the retry delay instead.

reaperrr commented:
Updated.
The `failCount` is now only reset after a certain amount of time if the number of buildings changed, otherwise `failRetryTicks` is reset instead.

This prevents the AI from performing 3 failing production attempts every minute or so in the situation where the failures were indeed caused by lack of space, and no buildings have been sold or destroyed in the meantime (and no construction yard added, either).

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +43,59 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 			playerPower = pm;
 			playerResources = pr;
 			this.category = category;
+			cachedBases = 0;
```
0 / false are the default values for the data types anyway, so these are redundant.

reaperrr committed:
Added random factor to building production delay

Early game AI usually follows the same build order (power plant first, then refinery), which also means they all start producing them at the same tick. This adds a random factor to the production delay, so not all AIs produce on the same tick.

Mailaender commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -130,22 +195,31 @@ ActorInfo GetProducibleBuilding(string commonName, IEnumerable<ActorInfo> builda
+			if ((actorInfo.Traits.WithInterface<PowerInfo>()
+				.Where(i => i.UpgradeMinEnabledLevel < 1).Sum(p => p.Amount) + playerPower.ExcessPower)
+				>= ai.Info.MinimumExcessPower)
+				return true;
```
You can simplify this to 
```csharp
return actorInfo.Traits.WithInterface<PowerInfo>().Where(i => i.UpgradeMinEnabledLevel < 1)
	.Sum(p => p.Amount) + playerPower.ExcessPower) >= ai.Info.MinimumExcessPower)
```

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +43,54 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+					.Where(a => a.Actor.Owner == player)
+					.Count();
+
+				var baseProviders = world.ActorsWithTrait<BaseBuilding>()
```
`BaseBuilding` (a tag trait used for the 'go to base' hotkey) != `BaseProvider` (the trait that grants base area)!

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +43,54 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+
+			if (!waterAvailable)
+			{
+				var currentBases = world.ActorsWithTrait<BaseBuilding>()
```
Similarly here.

pchote replied:
It looks like this query is run every tick, which is a bit overkill.  Can we rate limit it?

pchote commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +43,54 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 
+			if (!checkedWater)
+			{
+				waterAvailable = ai.EnoughWaterToBuildNaval();
```
Managing the state of two flags feels kind of gross.  Can we use an enum with three values instead?

pchote commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -231,6 +246,58 @@ public void Activate(Player p)
+		{
+			var baseBuildings = World.Actors.Where(
+				a => a.Owner == Player
+					&& a.HasTrait<BaseBuilding>()
```
`BaseProvider`.

reaperrr committed:
Improved naval placement check

Moved water checks before --waitTicks.
Use Water enum instead of multiple booleans.
Check for BaseProvider rather than BaseBuilding.
Move expensive ClosEnoughToWater check to last position for naval
production override.

obrakmann commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +51,57 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+					checkForBasesTicks = ai.Info.CheckForNewBasesDelay;
+			}
+
+			if (waterState == Water.NotEnoughWater && --checkForBasesTicks <= 0)
```
Due to the `if (--waitTicks > 0) return;` above, this counter will not decrease every tick, delaying re-checking considerably.

obrakmann commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -37,14 +51,57 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
+					waterState = Water.EnoughWater;
+				else
+					waterState = Water.NotEnoughWater;
+					checkForBasesTicks = ai.Info.CheckForNewBasesDelay;
```
missing braces on the `else` block

obrakmann commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -153,33 +234,69 @@ ActorInfo ChooseBuildingToBuild(ProductionQueue queue)
+			}
+
+			// Only consider building this if there is enough water inside the base perimeter and there are close enough adjacent buildings
+			if (waterState == Water.EnoughWater && ai.CloseEnoughToWater()
```
Can you move the `ai.CloseEnoughToWater()` call to the end of the conditions? It's the most costly.

obrakmann commented:
Just a remark: the AI would probably be much easier to program if it implemented all the INotify* interfaces, and all the events would then actually notify not only the actor, but its owner, too. Then we could get rid of this whole freaking state machine.

Phrohdoh commented:
re: @obrakmann's above comment; see [this](https://github.com/Phrohdoh/ModularAI).

obrakmann commented:
I think this should be ready once the inline comments are addressed. Looks good overall.

reaperrr commented:
Updated.

Decided to keep the checkForBasesTicks, and kept the commit separate since I didn't feel like wading through rebase conflicts due to the commits in between.

GraionDilach commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -52,12 +52,25 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
+		[Desc("A random delay (in ticks) of up to this is added to production delays.")]
+		public readonly int StructureProductionRandomBonusDelay = 10;
+
+		[Desc("How long to wait (in ticks) until retrying to build structure after the last 3 consecutive attempts failed.")]
```
You replaced "How long to wait" with "Minimum delay" above... consistency is for the weak? :D

reaperrr replied:
StructureProductionRandomBonusDelay is added to the production delays, but not this, hence why minimum for the above and not for these. So yes this is intentional.

GraionDilach replied:
This is `StructureProductionResumeDelay`'s description.

reaperrr replied:
Yes, and this does not add a random bonus delay, so it represents the exact delay value. The two Active and Inactive delays above add a random bonus delay, though, so they only represent the minimum delay. So the descriptions are correct, though it seems they can be misunderstood, so I'll try to word them better.

GraionDilach commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -52,12 +52,25 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
+		"for StructureProductionResumeDelay before retrying.")]
+		public readonly int MaximumFailedPlacementAttempts = 3;
+
+		[Desc("How long to wait (in ticks) until rechecking for new BaseProviders.")]
```
Again.

GraionDilach commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -82,6 +95,11 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
 		public readonly int MaxBaseRadius = 20;
 
+		[Desc("Radius in cells around each building with ProvideBuildableArea",
+		"to check for a 3x3 area of water where naval structures can be built.",
```
I'm not certain if I like the sound of this. RA2 naval yards are definitely bigger than 3x3 and IIRC that GDI navalyard c'n'p for TS is also 4x3. Can't you customize this that instead of a fixed size, it checks for the biggest foundation of the buildings listed in NavalProduction?

penev92 replied:
While I agree this is bad (too RA-specific), IMO we should limit the scope creeping already.
Also as far as I see the code that uses this radius doesn't really care about the footprint, so I'd say just remove the size from the description here.

reaperrr replied:
Factoring in bulding footprints (or generally more than 3x3 cells) would need another refactor, and I don't feel like doing that in this pr. This is already far better than what we had before.

GraionDilach replied:
@penev92: It explicitly does not check the footprint indeed, however it counts how many water tiles surrounded by fully buildable water tiles exist, which ultimately does end up with a 3x3 buildable water area.

@reaperrr: Please add TODOs into `EnoughWaterToBuildNaval()` and `CloseEnoughToWater()` then at least.

reaperrr committed:
Added TODOs for foundation of checked water cells

Fix description style issues

reaperrr commented:
Rebased & updated.

obrakmann commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -49,17 +49,32 @@ public sealed class HackyAIInfo : IBotInfo, ITraitInfo
 
-		[Desc("How long to wait (in ticks) between structure production checks when there is no active production.")]
+		[Desc("Delay (in ticks) between structure production checks when there is no active production.",
+		"A StructureProductionRandomBonusDelay is added to this.")]
```
Please remember to add spaces in front of a string continuation. Affects almost all `Desc`s here.

reaperrr replied:
Well, I misunderstood your comment anyway and tabbed lines 2+ on the descriptions instead ;)

obrakmann commented:
Aside from that style nit, this is looking good to me now. :+1:

reaperrr commented:
Updated.

Mailaender commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -250,6 +270,60 @@ public void Activate(Player p)
+				var playerWorld = Player.World;
+				var adjacentWater = Map.FindTilesInCircle(a.Location, Info.CheckForWaterRadius)
+					.Where(c => playerWorld.Map.Contains(c)
+						&& playerWorld.Map.GetTerrainInfo(c).Type == "Water"
```
You can use `.IsWater` instead which at least is cheaper than a string comparison.

reaperrr commented:
Updated.

Mailaender commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -250,6 +270,60 @@ public void Activate(Player p)
+
+			foreach (var a in areaProviders)
+			{
+				// TODO: Unhardcode terrain type
```
Is this resolved now?

reaperrr replied:
No, the idea is that in the future this whole check could be generalized to check for terrain type(s) of the modders choice, so the TODO still stands.

Mailaender commented:
Tested a bot match on a map with small ponds and limited space to build which the AI seems to solve a bit better now. :white_check_mark: Only noticed one blackout due to not building a power plant first.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0cfe138c44851a619b8166a2a734cb223c4057dd...6c0ddced2783473302d62733e9503e3491670eec)

Mailaender commented on OpenRA.Mods.Common/AI/BaseBuilder.cs:
```diff
@@ -55,7 +114,11 @@ public void Tick()
-			waitTicks = active ? ai.Info.StructureProductionActiveDelay : ai.Info.StructureProductionInactiveDelay;
+			// Add a random factor so not every AI produces at the same tick early in the game.
+			// Minimum should not be negative as delays in HackyAI could be zero.
+			var randomFactor = world.SharedRandom.Next(0, ai.Info.StructureProductionRandomBonusDelay);
```
Don't use shared random here in future patches. See https://github.com/OpenRA/OpenRA/pull/8870.
